### ISSUE TITLE
BufferStore: Buffers are not released..?

### ISSUE THREAD

**Issue by AndreasAakesson** [hioa-cs/IncludeOS#465]
Something is wrong with the release of buffers.

**TCP Test**
```
   [ Buf avail ] 562
                [x] inet->buffers_available() < buffers_available
                [x] conn.read() == small
  [ Buf avail ] 560
                + Enabling IRQ 0, mask: 0xfbfa
[ IRQ manager ] Updated subscriptions: 0x401 irq: 0
       [ TEST ] BIG string (6000)
  [ Buf avail ] 556
                [x] conn.read() == big
  [ Buf avail ] 546
       [ TEST ] HUGE string (60000)
  [ Buf avail ] 542
                [x] conn.read() == huge
  [ Buf avail ] 457
       [ TEST ] Connection
                [x] tcp.activeConnections() > 0
                [x] conn.is_connected()
                [x] conn.is_writable()
                [x] conn.is_state(ESTABLISHED)
       [ TEST ] Active close
                [x] !conn->is_writable()
                [x] conn.is_state(FIN-WAIT-1)
                [x] conn.is_state(FIN-WAIT-2)
                [x] conn.is_state(TIME-WAIT)
       [ TEST ] Outgoing Connection (10.0.2.2:8085)
                [x] conn->read() == small
                [x] Connection closed by server
       [ TEST ] Outgoing Internet Connection (india.colorado.edu:13)
                [x] Resolved host
                [x] Connected
                [x] Received data
       [ TEST ] Started 3 x MSL timeout.
       [ TEST ] Verify release of resources
                [x] tcp.activeConnections() == 0
                [ ] inet->buffers_available() == buffers_available
[ Buffers available ] 440
# TEST DONE #
```

**UDP Test**
```
Buffers available: 566
UDP server listening to port 4242 
================================================================================
 IncludeOS v0.8.0-rc.1-127-g11a8d54-dirty
 +--> Running [ IncludeOS UDP test ]
<Packet> CONSTRUCT packet, buf @ 0x2f200a
     [ DHCPv4 ] IP ADDRESS:     10.0.2.15
     [ DHCPv4 ] SUBNET MASK:     255.255.255.0
     [ DHCPv4 ] LEASE TIME:     35879872 mins
     [ DHCPv4 ] GATEWAY:     10.0.2.2
     [ DHCPv4 ] DNS SERVER:     192.168.1.1
<Packet> CONSTRUCT packet, buf @ 0x3218f0
<Packet> DESTRUCT packet, buf @ 0x3218f0
<Packet> DESTRUCT packet, buf @ 0x2f200a
<Packet> CONSTRUCT packet, buf @ 0x2f25f0
     [ DHCPv4 ] Server acknowledged our request!
      [ Inet4 ] Reconfiguring network. New IP: 10.0.2.15
[ DHCPv4::On_config ] Config complete
<Packet> DESTRUCT packet, buf @ 0x2f25f0
<Packet> CONSTRUCT packet, buf @ 0x2f2bd6
Getting UDP data from 10.0.2.2: 49733: Release my buffers already !!!!!!

Buffers available: 564
<Packet> CONSTRUCT packet, buf @ 0x322aa2
<Packet> CONSTRUCT packet, buf @ 0x323088
<Packet> DESTRUCT packet, buf @ 0x323088
Buffers available: 563
                + Enabling IRQ 0, mask: 0xfbfa
[ IRQ manager ] Updated subscriptions: 0x401 irq: 0
<Packet> DESTRUCT packet, buf @ 0x2f2bd6
<Packet> CONSTRUCT packet, buf @ 0x2f31bc
<Packet> DESTRUCT packet, buf @ 0x322aa2
<Packet> DESTRUCT packet, buf @ 0x2f31bc
Buffers available (1s delay): 562
```

**Every constructed packet is being destructed**
```
// formatted log from a TCP test
<Packet> CONSTRUCT packet, buf @ 0x311bd6    <Packet> DESTRUCT packet, buf @ 0x311bd6
<Packet> CONSTRUCT packet, buf @ 0x3121bc    <Packet> DESTRUCT packet, buf @ 0x3121bc
<Packet> CONSTRUCT packet, buf @ 0x3127a2    <Packet> DESTRUCT packet, buf @ 0x3127a2
<Packet> CONSTRUCT packet, buf @ 0x312d88    <Packet> DESTRUCT packet, buf @ 0x312d88
<Packet> CONSTRUCT packet, buf @ 0x31336e    <Packet> DESTRUCT packet, buf @ 0x31336e
<Packet> CONSTRUCT packet, buf @ 0x313954    <Packet> DESTRUCT packet, buf @ 0x313954
<Packet> CONSTRUCT packet, buf @ 0x313f3a    <Packet> DESTRUCT packet, buf @ 0x313f3a
<Packet> CONSTRUCT packet, buf @ 0x341aa2    <Packet> DESTRUCT packet, buf @ 0x341aa2
<Packet> CONSTRUCT packet, buf @ 0x342088    <Packet> DESTRUCT packet, buf @ 0x342088
<Packet> CONSTRUCT packet, buf @ 0x34323a    <Packet> DESTRUCT packet, buf @ 0x34323a
<Packet> CONSTRUCT packet, buf @ 0x3443ec    <Packet> DESTRUCT packet, buf @ 0x3443ec
<Packet> CONSTRUCT packet, buf @ 0x3449d2    <Packet> DESTRUCT packet, buf @ 0x3449d2
<Packet> CONSTRUCT packet, buf @ 0x345b84    <Packet> DESTRUCT packet, buf @ 0x345b84
```

**Some other output from TCP test**
```
       [ TEST ] HUGE string (60000)
<BufferStore::release_offset_buffer> Releasing 0x352410. 541 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x3529f6. 541 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x352fdc. 541 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x3535c2. 541 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x353ba8. 541 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x35418e. 541 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x354774. 541 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x354d5a. 541 available buffers.
...
// Incoming data can be a reason why the available buffers are decreasing.
  [ Buf avail ] 541
<BufferStore::release_offset_buffer> Releasing 0x362d7e. 539 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x36394a. 538 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x364afc. 536 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x36d2a6. 514 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x374e84. 494 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x375a50. 493 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x37661c. 492 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x3771e8. 491 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x377db4. 490 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x378980. 489 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x37954c. 488 available buffers.
<BufferStore::release_offset_buffer> Releasing 0x37a118. 487 available buffers.
```

**Comment by alfred-bratterud** [comment]
@AndreasAakesson if this is a bufstore problem it should be addressed in the [./test/bufstore](https://github.com/hioa-cs/IncludeOS/blob/dev/test/bufstore/service.cpp) test. Please review that test and update it if you find issues that it does not pick up

**Comment by AndreasAakesson** [comment]
This is fixed in #469
